### ISSUE TITLE
Polygon input in Object detection

### ISSUE THREAD

**Issue by thiendt2k1** [ultralytics/yolov5#12111]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, 
I would like to ask what happen to segmentation coordinates label when input into YOLOv5, as i thought the model only take 
class x_center y_center width height as input, but when i input a txt file contain all input as demo below, it is still work.
The file is a sample from Roboflow dataset: https://universe.roboflow.com/coffee-leaf-detection/rocole-xot5m/dataset/11
I trained on Colab V100, with config like below, on my custom dataset above.
`!python train.py --img 640 --batch-size 20 --weights yolov5l.pt --epochs 20 --data /content/yolov5/RoCoLe-11/data.yaml --name yolov5l `
**My Question:** 
1/ Is the label is convert to bounding box or what, if so, can you show me what is that function? 
2/ If not, how is the IoU is calculated when the input shape is polygon and the output is bounding boxes 
Thanks you.

### Additional
The sample txt file is like this: 
`3 0.36328125 0.725 0.42734375 0.683333334375 0.465625 0.6736111109375 0.49765625 0.715277778125 0.5484375 0.6986111109375 0.58359375 0.654166665625 0.6046875 0.6777777781250001 0.65078125 0.683333334375 0.66875 0.6375 0.70546875 0.6361111109375 0.77578125 0.6888888890625 0.85546875 0.6513888890625 0.93359375 0.577777778125 0.884375 0.5319444453125 0.85 0.5263888890625 0.8265625 0.47083333437499997 0.784375 0.4236111109375 0.75078125 0.4319444453125 0.72265625 0.383333334375 0.68046875 0.3638888890625 0.6453125 0.3361111109375 0.61015625 0.3513888890625 0.56484375 0.3236111109375 0.51484375 0.3138888890625 0.475 0.3180555546875 0.4390625 0.2805555546875 0.40703125 0.2805555546875 0.36328125 0.316666665625 0.31953125 0.3 0.28828125 0.34027777812500004 0.2375 0.345833334375 0.18359375 0.3805555546875 0.128125 0.39027777812500003 0.04296875 0.4375 0.0484375 0.477777778125 0.0828125 0.5361111109375 0.16015625 0.6083333343749999 0.1953125 0.6555555546875 0.259375 0.645833334375 0.30078125 0.6736111109375 0.36328125 0.725`
Here is a screenshot of a sample in roboflow:
![image](https://github.com/ultralytics/yolov5/assets/58879301/648a2c1f-887b-454b-a890-769fdb9ec456)

<details>

<summary>Data in roboflow</summary>
{
    "boxes": [
        {
            "type": "polygon",
            "label": "rust_level_2",
            "x": 312.5,
            "y": 321.7777775,
            "width": 570,
            "height": 284.444445,
            "points": [
                [
                    232.5,
                    464
                ],
                [
                    273.5,
                    437.333334
                ],
                [
                    298,
                    431.111111
                ],
                [
                    318.5,
                    457.777778
                ],
                [
                    351,
                    447.111111
                ],
                [
                    373.5,
                    418.666666
                ],
                [
                    387,
                    433.777778
                ],
                [
                    416.5,
                    437.333334
                ],
                [
                    428,
                    408
                ],
                [
                    451.5,
                    407.111111
                ],
                [
                    496.5,
                    440.888889
                ],
                [
                    547.5,
                    416.888889
                ],
                [
                    597.5,
                    369.777778
                ],
                [
                    566,
                    340.444445
                ],
                [
                    544,
                    336.888889
                ],
                [
                    529,
                    301.333334
                ],
                [
                    502,
                    271.111111
                ],
                [
                    480.5,
                    276.444445
                ],
                [
                    462.5,
                    245.333334
                ],
                [
                    435.5,
                    232.888889
                ],
                [
                    413,
                    215.111111
                ],
                [
                    390.5,
                    224.888889
                ],
                [
                    361.5,
                    207.111111
                ],
                [
                    329.5,
                    200.888889
                ],
                [
                    304,
                    203.555555
                ],
                [
                    281,
                    179.555555
                ],
                [
                    260.5,
                    179.555555
                ],
                [
                    232.5,
                    202.666666
                ],
                [
                    204.5,
                    192
                ],
                [
                    184.5,
                    217.777778
                ],
                [
                    152,
                    221.333334
                ],
                [
                    117.5,
                    243.555555
                ],
                [
                    82,
                    249.777778
                ],
                [
                    27.5,
                    280
                ],
                [
                    31,
                    305.777778
                ],
                [
                    53,
                    343.111111
                ],
                [
                    102.5,
                    389.333334
                ],
                [
                    125,
                    419.555555
                ],
                [
                    166,
                    413.333334
                ],
                [
                    192.5,
                    431.111111
                ],
                [
                    232.5,
                    464
                ]
            ]
        }
    ],
    "height": 640,
    "key": "C6P20E2_jpg.rf.84dd13aea235dc5a0a2c343f3c2cbc21.jpg",
    "width": 640
}
</details>

**Comment by glenn-jocher** [comment]
@thiendt2k1 hello,

In YOLOv5, the model expects the input labels to be in the format of class x_center y_center width height. However, it can still work with polygon coordinates if the points are converted to a bounding box.

To convert polygon coordinates to a bounding box, you can use functions from the `shapely` package in Python. Here is an example of how you can convert the polygon points to a bounding box:

```python
from shapely.geometry import Polygon

# Example polygon points
polygon_points = [[232.5, 464], [273.5, 437.333334], ...]

# Create a shapely polygon object
polygon = Polygon(polygon_points)

# Get the bounding box coordinates
x_min, y_min, x_max, y_max = polygon.bounds

# Convert to class x_center y_center width height format
width = x_max - x_min
height = y_max - y_min
x_center = x_min + width / 2
y_center = y_min + height / 2

# Normalize the coordinates
image_width = 640
image_height = 640
x_center /= image_width
y_center /= image_height
width /= image_width
height /= image_height

# Resulting bounding box coordinates in class x_center y_center width height format
result = f"{class_label} {x_center} {y_center} {width} {height}"
```

Regarding your second question about how IoU is calculated when the input shape is a polygon and the output shape is a bounding box, the IoU is calculated based on the intersection and union of the bounding box with the bounding box representation of the polygon. The polygon is converted to a bounding box, and then the IoU is calculated using the standard method.

I hope this helps! Let me know if you have any further questions.

**Comment by thiendt2k1** [comment]
@glenn-jocher 
Hi, thanks for the lightning reply, but i mean how the label is convert in yolov5, while wait for your reply, i poke around the code and find the function segments2boxes() in dataloader.py which is defined in general.py, it goes like this, 
`def segments2boxes(segments): 
    # Convert segment labels to box labels, i.e. (cls, xy1, xy2, ...) to (cls, xywh) 
    boxes = [] 
    for s in segments: 
        x, y = s.T  # segment xy
        boxes.append([x.min(), y.min(), x.max(), y.max()])  # cls, xyxy
    return xyxy2xywh(np.array(boxes))  # cls, xywh`

If i not mistaken then this will read the segment label, find the outer most x and y of the polygon like in your example right?

**Comment by glenn-jocher** [comment]
@thiendt2k1 Yes, you are exactly right. The `segments2boxes` function in YOLOv5's `dataloader.py` indeed converts segment labels to box labels. It does so by finding the outermost x and y coordinates of the polygon and then converting those coordinates to the (x, y, width, height) format required for YOLOv5's bounding box input.

I'm glad to hear that you found the function and its purpose. If you have any more questions or need further assistance, feel free to ask!
